cmd/snap/cmd_download.go

```diff
@@ -97,6 +127,189 @@ func fetchSnapAssertions(tsto *image.ToolingStore, snapPath string, snapInfo *sn
 	return assertPath, err
 }
 
+//
+type downloadStore interface {
+	DownloadSnap(name string, opts image.DownloadOptions) (targetFn string, info *snap.Info, err error)
+	AssertionFetcher(db *asserts.Database, save func(asserts.Assertion) error) asserts.Fetcher
+}
+
+var newDownloadStore = func() (downloadStore, error) {
+	return image.NewToolingStore()
+}
+
+func (x *cmdDownload) downloadDirect(snapName string, revision snap.Revision) error {
+	tsto, err := newDownloadStore()
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(Stdout, i18n.G("Fetching snap %q\n"), snapName)
+	dlOpts := image.DownloadOptions{
+		TargetDir: x.TargetDir,
+		Basename:  x.Basename,
+		Channel:   x.Channel,
+		CohortKey: x.CohortKey,
+		Revision:  revision,
+		// if something goes wrong, don't force it to start over again
+		LeavePartialOnError: true,
+	}
+	snapPath, snapInfo, err := tsto.DownloadSnap(snapName, dlOpts)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(Stdout, i18n.G("Fetching assertions for %q\n"), snapName)
+	assertPath, err := fetchSnapAssertionsDirect(tsto, snapPath, snapInfo)
+	if err != nil {
+		return err
+	}
+	printInstallHint(assertPath, snapPath)
+	return nil
+}
+
+func downloadSnapFromStreamWithProgress(downloadPath string, stream io.ReadCloser, dlInfo *client.DownloadInfo) error {
+	// TODO: support resume of exiting files and not download if the
+	//       file is already there with the right hash
+	f, err := os.OpenFile(downloadPath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// TODO: we have similar code like this in ToolingStore.DownloadSnap
+	// and store.downloadImpl
+
+	pbar := progress.MakeProgressBar()
+	pbar.Start(filepath.Base(downloadPath), float64(dlInfo.Size))
+	defer pbar.Finished()
+
+	// Intercept sigint
+	c := make(chan os.Signal, 3)
+	signal.Notify(c, syscall.SIGINT)
+	go func() {
+		<-c
```

Comment:
Isn't this leaking a goroutine here? Looked briefly at `signal.Reset()` but I didn't see any code that would close the channel passed to `signal.Notify()`.

Reply:
the goroutine will continue to run until the program gets a SIGINT or it exits, which is what we want, so I think it's ok in this case.

Reply:
Also the way of using Close() to break io.Copy() can be a bit awkward, and leads to this:
```
$ ./test-snap download core
Fetching snap "core"
core_7917.snap      0%  206kB/s 7m33s
### ^C
error: read unix @->/run/snapd.socket: use of closed network connection
```

Tweaked this a little:
```
diff --git a/cmd/snap/cmd_download.go b/cmd/snap/cmd_download.go
index c40ae7bc83..93fe7c1837 100644
--- a/cmd/snap/cmd_download.go
+++ b/cmd/snap/cmd_download.go
@@ -184,18 +184,28 @@ func downloadSnapFromStreamWithProgress(downloadPath string, stream io.ReadClose
        defer pbar.Finished()
 
        // Intercept sigint
-       c := make(chan os.Signal, 3)
+       c := make(chan os.Signal, 1)
        signal.Notify(c, syscall.SIGINT)
+       done := make(chan struct{}, 1)
+       canceled := false
        go func() {
-               <-c
-               pbar.Finished()
-               stream.Close()
+               select {
+               case <-c:
+                       canceled = true
+                       pbar.Finished()
+                       stream.Close()
+               case <-done:
+               }
        }()
+       defer close(done)
        defer signal.Reset(syscall.SIGINT)
 
        h := crypto.SHA3_384.New()
        mw := io.MultiWriter(f, h, pbar)
        if _, err := io.Copy(mw, stream); err != nil {
+               if canceled {
+                       return fmt.Errorf("download canceled")
+               }
                return err
        }
 
```

Now we get a nicer message:
```
$ snap download core
Fetching snap "core"
core_7917.snap     1%  602kB/s 2m34s
### ^C
error: download canceled
```